lib/system/web/src/keyboard_engine.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::{HashMap, BTreeSet};
+use std::cell::RefCell;
+use std::rc::Rc;
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+use itertools::Itertools;
+
+enum Action {
+    KeyUp,
+    KeyDown
+}
+
+type Bindings = HashMap<String, Box<dyn FnMut()>>;
+
+#[wasm_bindgen]
+pub struct KeyboardEngine {
+    bindings:        Rc<RefCell<Bindings>>,
+    _element:         web_sys::Element,
+    _prevent_default: bool,
+    //keydown_cb:    Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    //keyup_cb:      Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    _state:          Rc<RefCell<BTreeSet<String>>>
+}
+
+impl KeyboardEngine {
+    pub fn new(element: &web_sys::Element, prevent_default: bool) -> Self {
+        let state = Rc::new(RefCell::new(BTreeSet::new()));
+        let bindings = Rc::new(RefCell::new(HashMap::new()));
+        //let (keydown_cb, keyup_cb) = KeyboardEngine::init(element, bindings.clone(), state.clone());
+        KeyboardEngine::init(element, bindings.clone(), state.clone(), prevent_default);
+
+        Self {
+            bindings,
+            _element: element.clone(),
+            _prevent_default: prevent_default,
+            _state: state,
+
+            //keydown_cb: keydown_cb,
+            //keyup_cb: keyup_cb,
+        }
+    }
+
+    pub fn from_tag_name(html_tag: &str, prevent_default: bool) -> Self {
+        let window = web_sys::window().expect("no global window exists");
+        let document = window.document().expect("should have document on window");
+        let element = document.get_elements_by_tag_name(html_tag).item(0).expect("there is no such element");
+
+        KeyboardEngine::new(&element, prevent_default)
+    }
+
+    fn init (
+        element:         &web_sys::Element,
+        bindings:        Rc<RefCell<Bindings>>,
+        state:           Rc<RefCell<BTreeSet<String>>>,
+        prevent_default: bool
+    //) -> (
+    //    Closure<dyn FnMut(web_sys::KeyboardEvent)>,
+    //    Closure<dyn FnMut(web_sys::KeyboardEvent)>
+    ) {
+        let capture_key_cb = Rc::new(Box::new(move |mut code: String, action: Action| {
+            code.make_ascii_lowercase();
+            match action {
+                Action::KeyDown => {
+                    state.borrow_mut().insert(code);
+                    let key = state
+                        .borrow()
+                        .iter()
+                        .join("-");
+
+                    if let Some(f) = bindings.borrow_mut().get_mut(&key) { f(); }
```

Comment:
I don't really like the idea to searching for keys using strings. Sure, `Ctrl-A` is a nice visual way to show users the meaning of the key, but this should only be pretty printing utility. Comparisons should be done using bitfields and masks.